libvmi/driver/xen/xen.c

```diff
@@ -774,6 +774,73 @@ xen_get_tsc_info(
     return VMI_SUCCESS;
 }
 
+status_t
+xen_get_xsave_info(
+    vmi_instance_t vmi,
+    unsigned long vcpu,
+    xsave_area_t *xsave_info)
+{
+    xen_instance_t *xen = xen_get_instance(vmi);
+    uint32_t size = 0;
+    uint8_t *buf = NULL;
+    status_t ret = VMI_SUCCESS;
+    uint32_t off = 0;
+    struct hvm_hw_cpu_xsave *info = NULL;
+    struct hvm_save_descriptor *desc = NULL;
+
+    /* calling with no arguments --> return is the size of buffer required
+     *  for storing the HVM context
+     */
+    size = xen->libxcw.xc_domain_hvm_getcontext(xen->xchandle,xen->domainid, 0, 0);
+
+    if (size <= 0) {
+        errprint("Failed to fetch HVM context buffer size.\n");
+        ret = VMI_FAILURE;
+        goto _bail;
+    }
+
+    buf = malloc(size);
+    if (buf == NULL) {
+        errprint("Failed to allocate HVM context buffer.\n");
+        ret = VMI_FAILURE;
+        goto _bail;
+    }
+
+    if (xen->libxcw.xc_domain_hvm_getcontext(xen->xchandle,
+            xen->domainid, buf, size) < 0) {
+        errprint("Failed to fetch HVM context buffer.\n");
+        ret = VMI_FAILURE;
+        goto _bail;
+    }
+
+    off = 0;
+    while (off < size) {
+        desc = (struct hvm_save_descriptor *)(buf + off);
+
+        off += sizeof (struct hvm_save_descriptor);
+
+        if (desc->typecode == CPU_XSAVE_CODE && desc->instance == vcpu) {
+            info = (struct hvm_hw_cpu_xsave *)(buf + off);
+            break;
+        }
+        off += desc->length;
+    }
+
+    if (info == NULL) {
+        errprint("Failed to locate HVM xsave context.\n");
+        ret = VMI_FAILURE;
+        goto _bail;
+    }
+
+    memcpy(xsave_info, &info->save_area, sizeof(struct xsave_area));
```

Comment:
This assumes that the structure layout of `xsave_info` is the same as it is of `info->save_area`. That is not necessarily the case. Please copy fields explicitly instead. Also, add a safety check that makes sure `xsave_info` is not NULL.

Reply:
There is a small problem with the structure in xen 4.6 in that the xcomp_bv field was in the reserverd portion and then in 4.12 it came to be a field. If I assign per field the I will loose xcomp_bv on 4.6.

If that is the way we should go then I will have to have different versions according to xen version.

Reply:
> If that is the way we should go then I will have to have different versions according to xen version.

Please do that otherwise this code is very fragile and could easily break in the future and make it extremely hard to debug.